src/Modules/homological-algebra.jl

```diff
@@ -1,4 +1,498 @@
+export depth, fitting_ideal, is_flat, is_regular_sequence, koszul_homology, non_flat_locus 
+
 ###############################################################################
 # functionality supporting computations in homological algebra.
 ###############################################################################
 
+
+##############################################################################
+#
+# Fitting ideals
+#
+##############################################################################
+
+@doc Markdown.doc"""
+     fitting_ideal(M::ModuleFP, i::Int) 
+
+Return the `i`-th Fitting ideal of `M`.
+
+# Examples
+```jldoctest
+julia> R, (x, y) = PolynomialRing(QQ, ["x", "y"]);
+
+julia> F = free_module(R, 2);
+
+julia> o = zero(R)
+0
+
+julia> U = matrix([x^3-y^2 o; o x^3-y^2; -x^2 y; -y x])
+[x^3 - y^2           0]
+[        0   x^3 - y^2]
+[     -x^2           y]
+[       -y           x]
+
+julia> M = quo(F,U)[1]
+Subquotient of Submodule with 2 generators
+1 -> e[1]
+2 -> e[2]
+by Submodule with 4 generators
+1 -> (x^3 - y^2)*e[1]
+2 -> (x^3 - y^2)*e[2]
+3 -> -x^2*e[1] + y*e[2]
+4 -> -y*e[1] + x*e[2]
+
+julia> fitting_ideal(M, -1)
+ideal(0)
+
+julia> fitting_ideal(M, 0)
+ideal(x^3 - y^2)
+
+julia> fitting_ideal(M, 1)
+ideal(y, x)
+
+julia> fitting_ideal(M, 2)
+ideal(1)
+```
+"""
+function fitting_ideal(M::ModuleFP, i::Int)
+ error("not implemented for the given type of module.")
+end
+
+function fitting_ideal(F:: FreeMod, i::Int)
+ R = base_ring(F)
+ if !(coefficient_ring(R) isa AbstractAlgebra.Field)
+   throw(ArgumentError("The coefficient ring must be a field."))
+ end
+ if i < rank(F) return ideal(R, [zero(R)]) end
+ return R
+end
+
+function fitting_ideal(M::SubQuo, i::Int)
```

Comment:
```suggestion
function fitting_ideal(M::SubQuo{T}, i::Int) where {T<:MPolyElem}
```
Since the internals are using Singular functionality, this should probably be restricted to the polynomial case; just like the other methods below.